### ISSUE TITLE
Decision-focused learning pipeline

### ISSUE THREAD

**Issue by gdalle** [YasBroek/PRE_Internship#4]
We will focus on the following pipeline:

- Given a MAPF instance $x$, use a machine learning algorithm $\phi_w$ to predict modified edge weights $\theta$
- Given these modified edge weights, use a fast MAPF algorithm (like prioritized planning) to compute a solution $y$

The question is: how can we learn good parameters $w$ for the machine learning layer by gradient descent?

Sources:

- https://www.jair.org/index.php/jair/article/view/15320

Which methods from that paper can we apply? What kind of data do we have / do we need to generate?

**Comment by gdalle** [comment]
https://github.com/YasBroek/PRE_Internship/blob/208f5eb829adb212d297b5f5da4ebd93fc1be742/MAPF_code/src/TEG_methods.jl#L9-L27

It is likely that you're spending significant time in this function, because Graphs.jl has a coding flaw where `weights(g)` is called at the beginning of `astar` or `dijkstra`. For unweighted graphs or `SimpleWeightedGraph`s it is free so that doesn't matter, but here `weights(g)` for `TimeExpandedGraph` is expensive. It might be worth creating a `TimeExpandedWeights` struct that behaves like a matrix, using the same principles as for the graph.

https://github.com/YasBroek/PRE_Internship/blob/208f5eb829adb212d297b5f5da4ebd93fc1be742/MAPF_code/src/TEG_methods.jl#L97

Did you add unit tests for your inneighbor and outneighbor listing functions?

https://github.com/YasBroek/PRE_Internship/blob/208f5eb829adb212d297b5f5da4ebd93fc1be742/MAPF_code/src/find_path.jl#L15-L22

Is this heuristic valid regardless of the weights in your graph, or only for the original weights $1$ and $\sqrt{2}$?

https://github.com/YasBroek/PRE_Internship/blob/208f5eb829adb212d297b5f5da4ebd93fc1be742/MAPF_code/src/training.jl#L87

You want positive feature weights $w$ to ensure positive edge costs $c$ (assuming all your features $X$ are positive). Without positive edge costs, Dijkstra and A* are not correct and may not even terminate.
This might be the reason for the crash you observe.

Alternately, you could use a Generalized Linear Model instead of a standard Linear Regression: $c = \sigma(X w)$ instead of $c = X w$, where $\sigma$ is an activation function that ensures the output is positive, like a sigmoid.

Note that to preserve positive edge weights, you also need to take $\varepsilon$ small enough (this was actually a mistake in the paper and code by Berthet et al. 2020, which we corrected in my PhD).

https://github.com/YasBroek/PRE_Internship/blob/208f5eb829adb212d297b5f5da4ebd93fc1be742/MAPF_code/src/training.jl#L170

I don't understand why you initialize the sum to theta and not to zero.

https://github.com/YasBroek/PRE_Internship/blob/208f5eb829adb212d297b5f5da4ebd93fc1be742/MAPF_code/src/training.jl#L173-L175

If you use independent shortest paths for the gradient of the Fenchel-Young loss (which is the theoretically-motivated choice), you should also use them for computing the loss itself.

**Comment by YasBroek** [comment]
- [ ] Remember to adapt a_star heuristic to better represent graph with adapted weights
- [ ] Activation in training: try softplus and normalizing features
- [ ] Try applying projected Gradient Descent
- [ ] To use PerturbedMultiplicative -> install package from temporary branch -> InferOpt
        - [ ] If it doesn't work: TEG
- [ ] Set up latex doc for report -> sectioned structure, subsections, bullet points
- [ ] To think about:
        - [ ]  What are the scientific questions I want to contribute to?
        - [ ] What's different from prior studies? 
        - [ ] What are my contributions?
        - [ ] How do I back it up with experiments?

**Comment by gdalle** [comment]
- Don't initialize the weights to be negative, keep the projection instead of the softplus.
- Possibly: switch to vertex weights instead of edge weights (represented as a matrix instead of a flattened sparse vector).
- Play with parameters: gradient step size, perturbation scale $\varepsilon$.
- We want to see the cost decreasing as training progresses, and ideally a map with colors that make sense physically.
- Maybe the training data doesn't have enough information.
- Switch to Guillaume's package MultiAgentPathFinding.jl to speed things up
- Single instance, one scenario with the highest number of agents solvable with cooperative A* (Guillaume's code)
